slobrok/src/vespa/slobrok/server/rpcmirror.cpp

```diff
@@ -10,101 +10,67 @@ LOG_SETUP(".rpcmirror");
 namespace slobrok {
 
 IncrementalFetch::IncrementalFetch(FRT_Supervisor *orb,
-                         FRT_RPCRequest *req,
-                         VisibleMap &map,
-                         vespalib::GenCnt gen)
-    : FNET_Task(orb->GetScheduler()),
-      _req(req),
-      _map(map),
-      _gen(gen)
+                                   FRT_RPCRequest *req,
+                                   ServiceMapHistory &smh,
+                                   vespalib::GenCnt gen)
+  : FNET_Task(orb->GetScheduler()),
+    _req(req),
+    _smh(smh),
+    _gen(gen)
 { }
 
 IncrementalFetch::~IncrementalFetch() { }
 
-void
-IncrementalFetch::completeReq()
-{
-    vespalib::GenCnt newgen = _map.genCnt();
-    VisibleMap::MapDiff diff;
+void IncrementalFetch::completeReq(MapDiff diff) {
     FRT_Values &dst = *_req->GetReturn();
 
-    if (newgen == _gen) { // no change
-        dst.AddInt32(_gen.getAsInt());
-    } else if (_map.hasHistory(_gen)) {
-        diff = _map.history(_gen);
-        dst.AddInt32(_gen.getAsInt());
-    } else {
-        dst.AddInt32(0);
-        diff.updated = _map.allVisible();
-    }
+    dst.AddInt32(diff.fromGen.getAsInt());
 
     size_t sz = diff.removed.size();
-    FRT_StringValue *rem    = dst.AddStringArray(sz);
+    FRT_StringValue *rem = dst.AddStringArray(sz);
     for (uint32_t i = 0; i < sz; ++i) {
-        dst.SetString(&rem[i],  diff.removed[i].c_str());
+        dst.SetString(&rem[i], diff.removed[i].c_str());
     }
 
     sz = diff.updated.size();
-    FRT_StringValue *names  = dst.AddStringArray(sz);
-    FRT_StringValue *specs  = dst.AddStringArray(sz);
+    FRT_StringValue *names = dst.AddStringArray(sz);
+    FRT_StringValue *specs = dst.AddStringArray(sz);
     for (uint32_t i = 0; i < sz; ++i) {
-        dst.SetString(&names[i],  diff.updated[i]->getName().c_str());
-        dst.SetString(&specs[i],  diff.updated[i]->getSpec().c_str());
+        dst.SetString(&names[i],  diff.updated[i].name.c_str());
+        dst.SetString(&specs[i],  diff.updated[i].spec.c_str());
     }
 
-    dst.AddInt32(newgen.getAsInt());
+    dst.AddInt32(diff.toGen.getAsInt());
+
     LOG(debug, "mirrorFetch %p done (gen %d -> gen %d)",
-        this, _gen.getAsInt(), newgen.getAsInt());
+        this, diff.fromGen.getAsInt(), diff.toGen.getAsInt());
     _req->Return();
```

Comment:
consider setting _req to nullptr after returning it. The destructor could then verify appropriate destruction.